pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go

```diff
@@ -802,54 +807,86 @@ func (lrw *logicalReplicationWriterProcessor) flushBatch(
 //  2. The crd_internal_origin_timestamp requires modifying the user's schema.
 //
 // See the design document for possible solutions to both of these problems.
-func makeInsertQuery(fqTableName string, td catalog.TableDescriptor) string {
-	// TODO(ssd): Column families
-	var columnNames strings.Builder
-	var valueStrings strings.Builder
-	var onConflictUpdateClause strings.Builder
-	argIdx := 1
-	for _, col := range td.PublicColumns() {
-		// Virtual columns are not present in the data written to disk and
-		// thus not part of the rangefeed datum.
-		if col.IsVirtual() {
-			continue
+func makeInsertQueries(
+	fqTableName string, td catalog.TableDescriptor,
+) (map[catid.FamilyID]string, error) {
+	queries := make(map[catid.FamilyID]string, td.NumFamilies())
+
+	if err := td.ForeachFamily(func(family *descpb.ColumnFamilyDescriptor) error {
+		var columnNames strings.Builder
+		var valueStrings strings.Builder
+		var onConflictUpdateClause strings.Builder
+		argIdx := 1
+		seenIds := make(map[catid.ColumnID]struct{})
+		addColumn := func(colName string, colID catid.ColumnID) {
+			// We will set crdb_internal_origin_timestamp ourselves from the MVCC timestamp of the incoming datum.
+			// We should never see this on the rangefeed as a non-null value as that would imply we've looped data around.
+			if colName == "crdb_internal_origin_timestamp" {
+				return
+			}
+			if _, seen := seenIds[colID]; seen {
+				return
+			}
+
+			if argIdx == 1 {
+				columnNames.WriteString(colName)
+				fmt.Fprintf(&valueStrings, "$%d", argIdx)
+				fmt.Fprintf(&onConflictUpdateClause, "%s = $%d", colName, argIdx)
+			} else {
+				fmt.Fprintf(&columnNames, ", %s", colName)
+				fmt.Fprintf(&valueStrings, ", $%d", argIdx)
+				fmt.Fprintf(&onConflictUpdateClause, ",\n%s = $%d", colName, argIdx)
+			}
+			seenIds[colID] = struct{}{}
+			argIdx++
 		}
-		// We will set crdb_internal_origin_timestamp ourselves from the MVCC timestamp of the incoming datum.
-		// We should never see this on the rangefeed as a non-null value as that would imply we've looped data around.
-		if col.GetName() == "crdb_internal_origin_timestamp" {
-			continue
+
+		publicColumns := td.PublicColumns()
+		colOrd := catalog.ColumnIDToOrdinalMap(publicColumns)
+		primaryIndex := td.GetPrimaryIndex()
+
+		for i := 0; i < primaryIndex.NumKeyColumns(); i++ {
+			for i := 0; i < primaryIndex.NumKeyColumns(); i++ {
+				ord, ok := colOrd.Get(primaryIndex.GetKeyColumnID(i))
+				if !ok {
+					return errors.AssertionFailedf("expected to find column %d", ord)
+				}
+				col := publicColumns[ord]
```

Comment:
Well, I did it on purpose, but thanks for pointing it out because I did it based on a bad assumption about what was allowed in a primary index.  I'll add a test here.

Reply:
This should have probably been IsComputed

Reply:
I added a virtual and stored computed column to the test here. This will get easier to test once a different PR lands.

Reply:
> Virtual columns are allowed in primary indexes

i'm not going to ask further questions